File: ios/App/Repository/StatsRepo.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2021 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Combine

class StatsRepo: Startable {

    var blockedHot: AnyPublisher<Int, Never> {
        statsHot.map { it in Int(it.total_blocked) }.compactMap { $0 }
        .eraseToAnyPublisher()
    }

    var statsHot: AnyPublisher<CounterStats, Never> {
        writeStats.compactMap { $0 }.eraseToAnyPublisher()
    }

    private lazy var api = Services.apiForCurrentUser

    private lazy var accountIdHot = Repos.accountRepo.accountIdHot
    private lazy var enteredForegroundHot =  Repos.stageRepo.enteredForegroundHot
    private lazy var activityEntriesHot = Repos.activityRepo.entriesHot

    fileprivate let writeStats = CurrentValueSubject<CounterStats?, Never>(nil)

    fileprivate let refreshStatsT = SimpleTasker<Ignored>("refreshStats", debounce: 3.0)

    private var cancellables = Set<AnyCancellable>()

    func start() {
        onRefreshStats()
        onAccountIdChange_refreshStats()
        onForeground_refreshStats()
        onActivityChanged_refreshStats()
    }

    private func onRefreshStats() {
        refreshStatsT.setTask { _ in Just(true)
            .flatMap { _ in self.api.getStatsForCurrentUser() }
            .tryMap { it in self.writeStats.send(it) }
            .tryMap { _ in true }
            .eraseToAnyPublisher()
        }
    }

    // Refresh on account ID changed, will also trigger when starting the app
    private func onAccountIdChange_refreshStats() {
        accountIdHot
        .sink(onValue: { _ in self.refreshStatsT.send() })
        .store(in: &cancellables)
    }

    private func onForeground_refreshStats() {
        enteredForegroundHot
        .sink(onValue: { _ in self.refreshStatsT.send() })
        .store(in: &cancellables)
    }

    private func onActivityChanged_refreshStats() {
        activityEntriesHot
        .sink(onValue: { _ in self.refreshStatsT.send() })
        .store(in: &cancellables)
    }

}

File: android/app/src/main/java/repository/StatsRepo.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2022 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package repository

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import model.CounterStats
import service.Services
import utils.Ignored
import utils.SimpleTasker


class StatsRepo {

    private val writeStats = MutableStateFlow<CounterStats?>(null)

    val statsHot = writeStats.filterNotNull()
    val blockedHot = statsHot.map { it.total_blocked.toLong() }

    private val api = Services.apiForCurrentUser

    private val accountIdHot by lazy { Repos.account.accountIdHot }
    private val enteredForegroundHot by lazy { Repos.stage.enteredForegroundHot }
    private val activityEntriesHot by lazy { Repos.activity.entriesHot }

    private val refreshStatsT = SimpleTasker<Ignored>("refreshStats")

    fun start() {
        onRefreshStats()
        onAccountIdChange_refreshStats()
        onForeground_refreshStats()
        onActivityChanged_refreshStats()
    }

    private fun onRefreshStats() {
        refreshStatsT.setTask {
            val stats = api.getStatsForCurrentUser()
            writeStats.emit(stats)
            true
        }
    }

    // Refresh on account ID changed, will also trigger when starting the app
    private fun onAccountIdChange_refreshStats() {
        GlobalScope.launch {
            accountIdHot
            .collect {
                refreshStatsT.send()
            }
        }
    }

    private fun onForeground_refreshStats() {
        GlobalScope.launch {
            enteredForegroundHot
            .collect {
                refreshStatsT.send()
            }
        }
    }

    private fun onActivityChanged_refreshStats() {
        GlobalScope.launch {
            activityEntriesHot
            .collect {
                refreshStatsT.send()
            }
        }
    }

}